### PR TITLE
Add simulation support for G2/G3

### PR DESCRIPTION
**Pull request by FormerLurker**
This makes the simulator work at least, but it may need some modification that I'm not capable of doing without further study of your code.  There seems to be some functionality that is driving the extruder and trying to figure out how far behind it is.  The arc interpolation feature splits up a single gcode file line into many (all with the same line number) and adds the new segments to the buffer, but it looks like the timing is getting messed up a bit.  It's not bad, but is jerkier than I'd expect.  Perhaps you can take a look at that?  I'm actually not sure if that's expected behavior or not since I hadn't been running this prior to my modifications.

interpolateArc is now a direct member of the view model, as are the interpolation parameters. Also, I did have to change the interpolateArc code a bit to make it work in both cases, which probably made it somewhat less efficient.  It now clones the current state instead of directly modifying it.  Additionally, it now returns an array of points so they can be added outside the routine since the viewer and simulator have different methods and formats for adding new segments.  There is probably a better way to do this via some kind of function callback, but I'll leave that for another day if there are performance issues during the initial parsing/load.

There are probably a few finishing touches that could be added, but I believe the actual segment interpolation stuff is pretty good.

### CODE REVIEW

**Comment by FormerLurker** [comment]
Oh, and I've seen buffer overflows, again probably related to the routines expecting at most one segment per command.  Perhaps that can be dealt with as well?

**Comment by Kragrathea** [comment]
Cool! I will pull this and check it out as soon as I fix the bugs I am working on.

File: octoprint_prettygcode/static/js/prettygcode.js [hunk 1]
```diff
@@ -132,11 +331,49 @@ $(function () {
...
+                        // This is a g2/g3, so we need to do things a bit differently.
+                        // TODO: Do I need to convert axis mode here, or will things just work out?
+
+                        // is_clockwise
+
+                        // Extract I and J, R, and is_clockwise
+                        var is_clockwise = cmd.indexOf(" G2")>-1;
+                        var i = parseFloat(cmd.split("I")[1]);
+                        var j = parseFloat(cmd.split("J")[1]);
+                        var r = parseFloat(cmd.split("J")[1]);
```
Comment (Kragrathea) [inline octoprint_prettygcode/static/js/prettygcode.js:348]: Shouldn't this be cmd.split("R")?

**Comment by Kragrathea** [comment]
I am using a virtual printer for testing and there also seems to be an issue with the Z coordinate. At least while printing. 
With an arcwelded Benchy. If you watch the print head from the side after a few layers it will start ducking down around zero height for part of the layer. I caught it half way here:

image

File: octoprint_prettygcode/static/js/prettygcode.js (hunk 1 above)
Comment (FormerLurker) [reply octoprint_prettygcode/static/js/prettygcode.js:348]: Yes it should.  Didn't catch that since R isn't implemented.  Good catch

**Comment by FormerLurker** [comment]
Hmm..  if we can narrow down the line where this occurrs it should make it easier to debug.

**Comment by FormerLurker** [comment]
Should we create an issue for the Z-Height issue while printing arcs?  Might be a more convenient place to discuss and debug.

**Comment by Kragrathea** [comment]
Yes, I will create a new issue.

**Pull request merged**